smolyay/basis.py

```diff
@@ -335,17 +335,23 @@ def derivative(self, x):
         ValueError
             if input is outside the domain [-1,1].
         """
-        if x > 1 or x < -1:
+        if numpy.any([numpy.greater(x,1),numpy.less(x,-1)]):
             raise ValueError("Input is outside the domain [-1,1]")
-        if x == 1:
-            return self._n*(self._n+1)*(self._n+2)/3
-        elif x == -1:
-            return ((-1)**(self._n+1))*self._n*(self._n+1)*(self._n+2)/3
+        if self._derivative_polynomial is None:
+            self._derivative_polynomial = ChebyshevFirstKind(self._n+1)
+        x = numpy.array(x, copy=False)
+        y = numpy.zeros(x.shape)
+        flag1 = numpy.logical_not(numpy.logical_or(x == 1,x == -1))
+        flag2 = x == 1
+        flag3 = x == -1
```

Comment:
You should reuse these in the order I suggested so you don't duplicate calculations. Also, I think that it is allowed to use boolean operators, so I would suggest this as easier to read:
```suggestion
        flag2 = x == 1
        flag3 = x == -1
        flag1 = ~(flag2 | flag3)
```